components/metris/internal/gardener/handlers.go

```diff
@@ -0,0 +1,125 @@
+package gardener
+
+import (
+	"fmt"
+
+	gcorev1beta1 "github.com/gardener/gardener/pkg/apis/core/v1beta1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/client-go/tools/cache"
+)
+
+func (c *Controller) secretUpdateFunc(oldObj, newObj interface{}) {
+	newSecret, ok := newObj.(*corev1.Secret)
+	if !ok {
+		c.logger.Error("error decoding secret, invalid type")
+		return
+	}
+
+	oldSecret, ok := oldObj.(*corev1.Secret)
+	if !ok {
+		c.logger.Error("error decoding secret, invalid type")
+		return
+	}
+
+	if newSecret.ResourceVersion == oldSecret.ResourceVersion {
+		// Periodic resync will send update events for all known Secrets, so if the
+		// resource version did not change we skip
+		return
+	}
+
+	// check all shoots with that secret and update
+	fselector := fields.SelectorFromSet(fields.Set{fieldSecretBindingName: newSecret.Name}).String()
+
+	shootlist, err := c.gclientset.CoreV1beta1().Shoots(newSecret.Namespace).List(metav1.ListOptions{FieldSelector: fselector})
+	if err != nil || len(shootlist.Items) == 0 {
+		c.logger.Debugf("could not find shoots using secret '%s': skipping.", newSecret.Name)
+		return
+	}
+
+	for i := range shootlist.Items {
+		shoot := shootlist.Items[i]
+		c.logger.Debugf("secret '%s' has been modified, updating provider client config for shoot '%s'.", newSecret.Name, shoot.Name)
+		c.shootAddHandlerFunc(&shoot)
+	}
+}
+
+func (c *Controller) shootDeleteHandlerFunc(obj interface{}) {
+	var (
+		shoot *gcorev1beta1.Shoot
+		ok    bool
+	)
+
+	if shoot, ok = obj.(*gcorev1beta1.Shoot); !ok {
+		// try to recover deleted obj
+		delobj, ok := obj.(cache.DeletedFinalStateUnknown)
+		if !ok {
+			c.logger.Error("error decoding object, invalid type")
+			return
+		}
+
+		shoot, ok = delobj.Obj.(*gcorev1beta1.Shoot)
+		if !ok {
+			c.logger.Error("error decoding deleted object, invalid type")
+			return
+		}
+	}
+
+	shootname := fmt.Sprintf("%s--%s", shoot.Namespace, shoot.Name)
+
+	c.logger.Warnf("shoot '%s' is being remove from the provider accounts", shootname)
+
+	c.accountsChan <- &Account{Name: shootname}
+}
+
+func (c *Controller) shootAddHandlerFunc(obj interface{}) {
+	if shoot, ok := obj.(*gcorev1beta1.Shoot); ok {
+		pname := shoot.Spec.CloudProfileName
+		if pname == "az" {
+			pname = "azure"
+		}
+
+		if c.providertype == pname {
+			accountid := shoot.GetLabels()[labelAccountID]
+
+			subaccountid, ok := shoot.GetLabels()[labelSubAccountID]
+			if !ok || subaccountid == "" {
+				// if subaccount label is empty, try the annotation ?
+				subaccountid, ok = shoot.GetAnnotations()[annotationSubAccountID]
```

Comment:
SubAccount from the annotation is a SubAccount for audit logs not the one with which the provisioning was requested, therefore I am not sure if you can just use this one if the label is not present.